File: app/src/main/java/io/github/fvasco/pinpoi/util/DebugUtil.java
package io.github.fvasco.pinpoi.util;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import io.github.fvasco.pinpoi.BuildConfig;
import io.github.fvasco.pinpoi.dao.PlacemarkCollectionDao;
import io.github.fvasco.pinpoi.dao.PlacemarkDao;
import io.github.fvasco.pinpoi.model.Placemark;
import io.github.fvasco.pinpoi.model.PlacemarkAnnotation;
import io.github.fvasco.pinpoi.model.PlacemarkCollection;

/**
 * Debug utilities
 *
 * @author Francesco Vasco
 */
public final class DebugUtil {
    private DebugUtil() {
    }

    public static void setUpDebugDatabase(final Context context) {
        if (!BuildConfig.DEBUG) throw new Error();

        try (final PlacemarkCollectionDao placemarkCollectionDao = new PlacemarkCollectionDao(context).open();
             final PlacemarkDao placemarkDao = new PlacemarkDao(context).open()) {
            final SQLiteDatabase placemarkCollectionDatabase = placemarkCollectionDao.getDatabase();
            final SQLiteDatabase placemarkDatabase = placemarkDao.getDatabase();
            placemarkCollectionDatabase.beginTransaction();
            placemarkDatabase.beginTransaction();
            try {
                // clear all database
                for (final PlacemarkCollection placemarkCollection : placemarkCollectionDao.findAllPlacemarkCollection()) {
                    placemarkDao.deleteByCollectionId(placemarkCollection.getId());
                    placemarkCollectionDao.delete(placemarkCollection);
                }

                // recreate test database
                final PlacemarkCollection placemarkCollection = new PlacemarkCollection();
                for (int pci = 15; pci >= 0; --pci) {
                    placemarkCollection.setId(0);
                    placemarkCollection.setName("Placemark Collection " + pci);
                    placemarkCollection.setCategory(pci == 0 ? null : "Category " + (pci % 7));
                    placemarkCollection.setDescription(placemarkCollection.getName() + " long long long description");
                    placemarkCollection.setSource("http://source/" + pci + ".csv");
                    placemarkCollection.setPoiCount(pci);
                    placemarkCollection.setLastUpdate(pci * 10_000_000);
                    placemarkCollectionDao.insert(placemarkCollection);
                    Log.i(DebugUtil.class.getSimpleName(), "inserted " + placemarkCollection);

                    final Placemark placemark = new Placemark();
                    for (int lat = -60; lat < 60; ++lat) {
                        for (int lon = -90; lon < 90; lon += 2) {
                            placemark.setId(0);
                            placemark.setName("Placemark " + lat + "," + lon + "/" + pci);
                            placemark.setDescription((lat + lon) % 10 == 0 ? null : placemark.getName() + " description");
                            placemark.setLatitude((float) (lat + Math.sin(lat + pci)));
                            placemark.setLongitude((float) (lon + Math.sin(lon - pci)));
                            placemark.setCollectionId(placemarkCollection.getId());
                            placemarkDao.insert(placemark);

                            if ((lat + lon + pci) % 9 == 0) {
                                final PlacemarkAnnotation placemarkAnnotation = placemarkDao.loadPlacemarkAnnotation(placemark);
                                placemarkAnnotation.setFlagged((lat + lon + pci) % 3 == 0);
                                placemarkAnnotation.setNote("Placemark annotation for " + placemark.getName());
                                placemarkDao.update(placemarkAnnotation);
                            }
                        }
                    }
                }

                placemarkDatabase.setTransactionSuccessful();
                placemarkCollectionDatabase.setTransactionSuccessful();
            } finally {
                placemarkDatabase.endTransaction();
                placemarkCollectionDatabase.endTransaction();
            }
        }
    }


}

File: app/src/main/java/io/github/fvasco/pinpoi/util/DebugUtil.kt
package io.github.fvasco.pinpoi.util

import android.content.Context
import android.util.Log
import io.github.fvasco.pinpoi.BuildConfig
import io.github.fvasco.pinpoi.dao.PlacemarkCollectionDao
import io.github.fvasco.pinpoi.dao.PlacemarkDao
import io.github.fvasco.pinpoi.model.Placemark
import io.github.fvasco.pinpoi.model.PlacemarkCollection
import kotlin.math.sin

/**
 * Debug utilities

 * @author Francesco Vasco
 */
fun setUpDebugDatabase(context: Context) {
    if (!BuildConfig.DEBUG) throw AssertionError()

    val placemarkCollectionDao = PlacemarkCollectionDao(context)
    placemarkCollectionDao.open()
    val placemarkDao = PlacemarkDao(context)
    placemarkDao.open()
    try {
        val placemarkCollectionDatabase = placemarkCollectionDao.database!!
        val placemarkDatabase = placemarkDao.database!!
        placemarkCollectionDatabase.beginTransaction()
        placemarkDatabase.beginTransaction()
        try {
            // clear all database
            for (placemarkCollection in placemarkCollectionDao.findAllPlacemarkCollection()) {
                placemarkDao.deleteByCollectionId(placemarkCollection.id)
                placemarkCollectionDao.delete(placemarkCollection)
            }

            // recreate test database
            val placemarkCollection = PlacemarkCollection()
            for (pci in 0..15) {
                placemarkCollection.id = 0
                placemarkCollection.name = "Placemark Collection '$pci'"
                placemarkCollection.category = if (pci == 0) "" else "Category " + pci % 7
                placemarkCollection.description =
                    placemarkCollection.name + " long long long description"
                placemarkCollection.source = "http://www.example.org/poi-collection-$pci.csv"
                placemarkCollection.poiCount = pci
                placemarkCollection.lastUpdate = pci * 10000000L
                placemarkCollectionDao.insert(placemarkCollection)
                Log.i("setUpDebugDatabase", "inserted " + placemarkCollection)

                val placemark = Placemark()
                for (lat in -60..60) {
                    for (lon in -90..90 step 2) {
                        placemark.id = 0
                        placemark.name = "Placemark $lat,$lon / $pci"
                        placemark.description = when {
                            (lat + lon) % 10 == 0 -> ""
                            pci == 0 -> placemark.name + "<u>beautiful</u> description"
                            else -> placemark.name + " description"
                        }
                        placemark.coordinates = Coordinates(
                            (lat + sin((lat + pci).toDouble())).toFloat(),
                            (lon + sin((lon - pci).toDouble())).toFloat()
                        )
                        placemark.collectionId = placemarkCollection.id
                        placemarkDao.insert(placemark)

                        if ((lat + lon + pci) % 9 == 0) {
                            val placemarkAnnotation =
                                placemarkDao.loadPlacemarkAnnotation(placemark)
                            placemarkAnnotation.flagged = (lat + lon + pci) % 3 == 0
                            placemarkAnnotation.note = "Placemark annotation for " + placemark.name
                            placemarkDao.update(placemarkAnnotation)
                        }
                    }
                }
            }

            placemarkDatabase.setTransactionSuccessful()
            placemarkCollectionDatabase.setTransactionSuccessful()
        } finally {
            placemarkDatabase.endTransaction()
            placemarkCollectionDatabase.endTransaction()
        }
    } finally {
        placemarkDao.close()
        placemarkCollectionDao.close()
    }
}
